File: modules/04_privacy_preserving_ai/code_samples/android_derived_data/HeartRateFeatures.kt
data class HeartRateSample(val timestamp: Long, val bpm: Double)

data class HeartRateFeatures(
    val mean: Double,
    val min: Double,
    val max: Double,
    val stdDev: Double
)

fun computeFeatures(samples: List<HeartRateSample>): HeartRateFeatures {
    if (samples.isEmpty()) return HeartRateFeatures(0.0, 0.0, 0.0, 0.0)

    val values = samples.map { it.bpm }
    val mean = values.average()
    val min = values.minOrNull() ?: mean
    val max = values.maxOrNull() ?: mean
    val variance = values.map { (it - mean) * (it - mean) }.average()
    val stdDev = kotlin.math.sqrt(variance)

    return HeartRateFeatures(mean, min, max, stdDev)
}

File: modules/04_privacy_preserving_ai/code_samples/ios_derived_data/HeartRateFeatures.swift
struct HeartRateSample {
    let timestamp: Date
    let bpm: Double
}

struct HeartRateFeatures {
    let mean: Double
    let min: Double
    let max: Double
    let stdDev: Double
}

func computeFeatures(from samples: [HeartRateSample]) -> HeartRateFeatures {
    let values = samples.map { $0.bpm }
    guard !values.isEmpty else {
        return HeartRateFeatures(mean: 0, min: 0, max: 0, stdDev: 0)
    }

    let mean = values.reduce(0, +) / Double(values.count)
    let minV = values.min() ?? mean
    let maxV = values.max() ?? mean
    let variance = values
        .map { pow($0 - mean, 2) }
        .reduce(0, +) / Double(values.count)
    let std = sqrt(variance)

    return HeartRateFeatures(mean: mean, min: minV, max: maxV, stdDev: std)
}
